### PR TITLE
Vectorized NHLBI compression and decompression

### PR DESCRIPTION
**Pull request by mirounga**
Chooses the optimized version at runtime based on the available CPU instruction set.
Compression is available in legacy (scalar), SSE4.1, and AVX2 versions.
Decompression in scalar and AVX2.
Unit tests cover the concordance between the versions.

### CODE REVIEW

**Review by dchansen — COMMENTED** [review]
This is a nice feature to have. 
I think you have confused "inline" and "static" in the code. That needs to be fixed. Maybe you want static inline? I suspect inline will not actually do anything here.

I am not able to judge the vectorized code, but the tests seem reasonable.

File: gadgets/mri_core/cpuisa.cpp [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+#else
+// SIMD intrinsics for GCC
+#include <x86intrin.h>
+#include <cpuid.h>
+#define cpuid(info, level) __cpuid(level, info[0], info[1], info[2], info[3]);
+#define cpuidex(info, leaf, subleaf) __cpuid_count(leaf, subleaf, info[0], info[1], info[2], info[3]);
+#endif
+
+#include "cpuisa.h"
+
```
[the remark this answers is not in the document]
Comment (mirounga) [reply gadgets/mri_core/cpuisa.cpp:17]: Done

File: gadgets/mri_core/CompressedFloatBufferSse41.cpp [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+#include <immintrin.h>
+using namespace NHLBI;
+
+
+InstructionSet CompressedFloatBufferSse41::getInstructionSet()
+{
+    return InstructionSet::Sse41;
+}
+
+inline __m128i scale_and_cast_ps(const __m128 _float_val, const __m128 _scale, const __m128 _sign_mask, __m128i& _sign)
```
[the remark this answers is not in the document]
Comment (mirounga) [reply gadgets/mri_core/CompressedFloatBufferSse41.cpp]: Done

File: gadgets/mri_core/CompressedFloatBufferAvx2.cpp [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+
+using namespace NHLBI;
+
+InstructionSet CompressedFloatBufferAvx2::getInstructionSet()
+{
+    return InstructionSet::Avx2;
+}
+
+// Vector scale, round, and convert to integer
+inline __m256i scale_and_cast_ps(const __m256 _float_val, const __m256 _scale, const __m256 _sign_mask, __m256i& _sign)
```
[the remark this answers is not in the document]
Comment (mirounga) [reply gadgets/mri_core/CompressedFloatBufferAvx2.cpp]: done

**Review by dchansen — APPROVED** [review]

**Pull request merged**